modules/packages/ZMQ.chpl

```diff
@@ -0,0 +1,451 @@
+  private extern const ZMQ_MECHANISM: c_int;
+  private extern const ZMQ_PLAIN_SERVER: c_int;
+  private extern const ZMQ_PLAIN_USERNAME: c_int;
+  private extern const ZMQ_PLAIN_PASSWORD: c_int;
+  private extern const ZMQ_CURVE_SERVER: c_int;
+  private extern const ZMQ_CURVE_PUBLICKEY: c_int;
+  private extern const ZMQ_CURVE_SECRETKEY: c_int;
+  private extern const ZMQ_CURVE_SERVERKEY: c_int;
+  private extern const ZMQ_PROBE_ROUTER: c_int;
+  private extern const ZMQ_REQ_CORRELATE: c_int;
+  private extern const ZMQ_REQ_RELAXED: c_int;
+  private extern const ZMQ_CONFLATE: c_int;
+  private extern const ZMQ_ZAP_DOMAIN: c_int;
+
+  // -- Message Options
+  private extern const ZMQ_MORE: c_int;
+
+  // -- Send/Recv Options
+  private extern const ZMQ_DONTWAIT: c_int;
+  private extern const ZMQ_SNDMORE: c_int;
+
+  // -- Security Options
+  private extern const ZMQ_NULL: c_int;
+  private extern const ZMQ_PLAIN: c_int;
+  private extern const ZMQ_CURVE: c_int;
+
+  /*
+    Query the ZMQ library version.
+
+    :returns: An :type:`(int,int,int)` tuple of the major, minor, and patch
+        version of the ZMQ library.
+   */
+  proc version: (int,int,int) {
+    var major, minor, patch: c_int;
+    zmq_version(c_ptrTo(major), c_ptrTo(minor), c_ptrTo(patch));
+    return (major:int, minor:int, patch:int);
+  }
+
+  /*
+    ZMQ.Context
+
+    The ZMQ Context class maintains a vector of child sockets.  This allows
+    users to delete the Context object without managing the closure and
+    deletion of Sockets created from this Context.
+
+    In ZMQ's C API, context "objects" are thread safe. The lock internal to
+    this Chapel class is only used to serialize accesses to the vector-style
+    array of child sockets.
+  */
+  class Context {
+
+    // Member Elements
+    var ctx: c_void_ptr;
+    var socks: [1..0] Socket;
+    var lock: sync bool;
+
+    // Constructor
+    proc Context() {
+      this.ctx = zmq_ctx_new();
+      if this.ctx == nil {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in Context(): %s\n", errmsg);
+      }
+    }
+
+    // Destructor
+    pragma "no doc"
+    proc ~Context() {
+      // Close any open sockets
+      for sock in socks do
+        delete sock;
+      // Terminate the context
+      var ret = zmq_ctx_term(this.ctx):int;
+      if ret == -1 {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in ~Context(): %s\n", errmsg);
+      }
+    }
+
+    // socket()
+    proc socket(socktype: int, parSafe: bool = true,
+                linger: int = -1): Socket{
+      var sock = new Socket(this, socktype, parSafe, linger);
+      this.lock = true;
+      this.socks.push_back(sock);
+      this.lock;
+      return sock;
+    }
+
+  } // class Context
+
+  /*
+    ZMQ.Socket
+
+    Design Notes:
+
+    - Do not delete Socket objects; they are deleted when the parent
+      Context object is deleted, however, they can be closed early.
```

Comment:
I don't think it would be too difficult to handle explicit deletion of Socket objects. In fact, I think I added a Context reference inside to the Socket object for that purpose -- but never implemented it in Socket.close().